src/inc/til/rle.h

```diff
@@ -0,0 +1,1064 @@
+            // Quite ironic given that the committee refuses to change the STL ABI,
+            // forcing us to reinvent std::pair as til::rle_pair.
+            // --> Let's assume that container behavior falls into only two categories:
+            //     * Moves the underlying memory, setting .size() to 0
+            //     * Leaves the source intact (basically copying it)
+            //     We can detect these cases using _runs.empty() and set _total_length accordingly.
+            if (other._runs.empty())
+            {
+                other._total_length = 0;
+            }
+        }
+
+        basic_rle& operator=(basic_rle&& other) noexcept
+        {
+            _runs = std::move(other._runs);
+            _total_length = other._total_length;
+
+            // See basic_rle(basic_rle&&) for why this is necessary.
+            if (other._runs.empty())
+            {
+                other._total_length = 0;
+            }
+
+            return *this;
+        }
+
+        basic_rle(std::initializer_list<rle_type> runs) :
+            _runs(runs), _total_length(0)
+        {
+            for (const auto& run : _runs)
+            {
+                _total_length += run.length;
+            }
+        }
+
+        basic_rle(container&& runs) :
+            _runs(std::forward<container>(runs)), _total_length(0)
+        {
+            for (const auto& run : _runs)
+            {
+                _total_length += run.length;
+            }
+        }
+
+        basic_rle(const size_type length, const value_type& value) :
+            _total_length(length)
+        {
+            if (length)
+            {
+                _runs.emplace_back(value, length);
+            }
+        }
+
+        void swap(basic_rle& other) noexcept
+        {
+            _runs.swap(other._runs);
+            std::swap(_total_length, other._total_length);
+        }
+
+        bool empty() const noexcept
+        {
+            return _total_length == 0;
+        }
+
+        // Returns the total length of all runs as encoded.
+        size_type size() const noexcept
+        {
+            return _total_length;
+        }
+
+        // This method gives access to the raw run length encoded array
+        // and allows users of this class to iterate over those.
+        const container& runs() const noexcept
+        {
+            return _runs;
+        }
+
+        // Get the value at the position
+        const_reference at(size_type position) const
+        {
+            const auto begin = _runs.begin();
+            const auto end = _runs.end();
+
+            rle_scanner scanner(begin, end);
+            auto it = scanner.scan(position).first;
+
+            if (it == end)
+            {
+                throw std::out_of_range("position out of range");
+            }
+
+            return it->value;
+        }
+
+        [[nodiscard]] basic_rle slice(size_type start_index, size_type end_index) const noexcept

```

Comment:
Add comment explaining whether end_index is inclusive or exclusive and perhaps also mention that "slice" is sort of like "substring" for those more acquainted with strings.